operator/controllers/helmrelease_controller.go

```diff
@@ -115,7 +153,54 @@ func (r *HelmReleaseReconciler) onDelete(ctx context.Context, rls shipitv1beta1.
 	return errNotImplemented
 }
 
+func isHelmReleaseNotFound(name string, err error) bool {
+	// dynamic errors can't be directly compared for equality. We use the
+	// error message string, though there's no guarantee it won't change.
+	return err.Error() == helmerrors.ErrReleaseNotFound(name).Error()
+}
+
 func (r *HelmReleaseReconciler) onUpdate(ctx context.Context, rls shipitv1beta1.HelmRelease) (ctrl.Result, error) {
+	releaseName := rls.Spec.ReleaseName
+
+	_, err := r.helm.ReleaseStatus(releaseName)
+	if err != nil {
+		if isHelmReleaseNotFound(releaseName, err) {
+			return r.install(ctx, rls)
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get release status for %s", releaseName)
+	}
+
+	return ctrl.Result{}, errNotImplemented
+}
+
+func (r *HelmReleaseReconciler) install(ctx context.Context, rls shipitv1beta1.HelmRelease) (ctrl.Result, error) {
+	chartURI := rls.Spec.Chart.URI()
+	releaseName := rls.Spec.ReleaseName
+
+	chart, err := r.downloader.Download(ctx, chartURI)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to download chart %s", chartURI)
+	}
+
+	// TODO: use the returned response's `Release.Manifest` to watch and
+	// receive events for the k8s resources owned by this chart
+	if _, err := r.helm.InstallReleaseFromChart(chart, r.Namespace, helm.ReleaseName(releaseName)); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to install release %s using chart %s", releaseName, chartURI)
+	}
+
+	rls.Status.SetCondition(shipitv1beta1.HelmReleaseCondition{
+		Type:    release.Status_PENDING_INSTALL.String(),
+		Message: fmt.Sprint("installing chart", chartURI),
+	})
+
+	if err := r.Update(ctx, &rls); err != nil {
+		r.Log.Info("failed to update HelmRelease status", "release", releaseName, "status", release.Status_PENDING_INSTALL)
```

Comment:
There's a bit of chicken-egg here. Do we set the HelmRelease status to PENDING_INSTALL before calling helm (what if the helm call fails?) or do we call helm before setting PENDING_INSTALL (what if the HelmRelease update fails?)?

Reply:
We should be calling helm then setting `PENDING_INSTALL`. I don't think we need to be so discrete as to represent the state of the controller having become aware that it needs to run install but hasn't done so yet.